includes/gateways/paypal/ipn.php

```diff
@@ -0,0 +1,386 @@
+ *
+ * @package    easy-digital-downloads
+ * @subpackage Gateways\PayPal\Webhooks
+ * @copyright  Copyright (c) 2021, Sandhills Development, LLC
+ * @license    GPL2+
+ * @since      2.11
+ */
+
+namespace EDD\Gateways\PayPal\IPN;
+
+/**
+ * Listens for an IPN call from PayPal
+ *
+ * This is intended to be a 'backup' listener, for if the webhook is no longer connected for a specific PayPal object.
+ *
+ * @since 3.1.0.3
+ */
+function listen_for_ipn() {
+	if ( empty( $_GET['edd-listener'] ) || 'eppe' !== $_GET['edd-listener'] ) {
+		return;
+	}
+
+	ipn_debug_log( 'IPN Backup Loaded' );
+
+	nocache_headers();
+
+	$verified = false;
+
+	// Set initial post data to empty string.
+	$post_data = '';
+
+	// Fallback just in case post_max_size is lower than needed.
+	if ( ini_get( 'allow_url_fopen' ) ) {
+		$post_data = file_get_contents( 'php://input' );
+	} else {
+		// If allow_url_fopen is not enabled, then make sure that post_max_size is large enough.
+		ini_set( 'post_max_size', '12M' );
+	}
+
+	// Start the encoded data collection with notification command.
+	$encoded_data = 'cmd=_notify-validate';
+
+	// Get current arg separator.
+	$arg_separator = edd_get_php_arg_separator_output();
+
+	// Verify there is a post_data.
+	if ( $post_data || strlen( $post_data ) > 0 ) {
+
+		// Append the data.
+		$encoded_data .= $arg_separator . $post_data;
+
+	} else {
+
+		// Check if POST is empty.
+		if ( empty( $_POST ) ) {
+
+			// Nothing to do.
+			ipn_debug_log( 'post data not detected, bailing' );
+			return;
+
+		} else {
+
+			// Loop through each POST.
+			foreach ( $_POST as $key => $value ) {
+
+				// Encode the value and append the data.
+				$encoded_data .= $arg_separator . "$key=" . urlencode( $value );
+
+			}
+
+		}
+
+	}
+
+	// Convert collected post data to an array.
+	parse_str( $encoded_data, $encoded_data_array );
+
+	// We're always going to validate the IPN here...
+	if ( ! edd_is_test_mode() ) {
+
+		ipn_debug_log( 'preparing to verify IPN data' );
+
+		// Validate the IPN.
+		$remote_post_vars = array(
+			'method'      => 'POST',
+			'timeout'     => 45,
+			'redirection' => 5,
+			'httpversion' => '1.1',
+			'blocking'    => true,
+			'body'        => $encoded_data_array,
+			'headers'     => array(
+				'host'         => 'www.paypal.com',
+				'connection'   => 'close',
+				'content-type' => 'application/x-www-form-urlencoded',
+				'post'         => '/cgi-bin/webscr HTTP/1.1',
+
+			),
+		);
+
+		// Get response.
+		$api_response = wp_remote_post( edd_get_paypal_redirect(), $remote_post_vars );
+		$body         = wp_remote_retrieve_body( $api_response );
+
+		if ( is_wp_error( $api_response ) ) {
+			/* Translators: %s - IPN Verification response */
+			edd_record_gateway_error( __( 'IPN Error', 'easy-digital-downloads' ), sprintf( __( 'Invalid PayPal Express IPN verification response. IPN data: %s', 'easy-digital-downloads' ), json_encode( $api_response ) ) );
```

Comment:
Should this mention Express?

Related: I'm confused by the early return if the listener is not `eppe`, which I thought was PayPal Express.

Reply:
@robincornett Most people, at this point who are on commerce with subscriptions were on PayPal Pro/Express. They only get one IPN URL, and we need their Pro/Express subs to still come through, so we're going to use the same endpoint and just rely on the gateway check.